Version 3.14.0 proposal

tim-schilling opened the pull request:
Currently projects using Django 4.1 are blocked from using Django Rest Framework due to #8591. Since that was merged, I think a release is warranted.

I started putting together the changes from 3.13.1, and discovered there were a couple of API changes. I suspect this should require a minor version change, but a maintainer should make that final call.

If the version change isn't desired, let me know and I can remove that.

tomchristie commented:
Good stuff, thanks Tim.
And yes, a 3.14 release makes sense.

I'd like #7522 to be part of this too, but would be great to get a review onto that from someone other than myself.

adamchainz commented on docs/community/release-notes.md:
```diff
@@ -34,6 +34,23 @@ You can determine your currently installed version using `pip show`:
+
+Date: 10th August 2022
+
+* Enforce `is_valid(raise_exception=False)` as a keyword-only argument. [#7952]
+* Django 4.1 compatability. [#8591]
+* Stop calling `set_context` on Validators. [#8589]
+* Return `NotImplemented` from `ErrorDetails.__ne__`. [#8538]
+* Don't evaluate `DateTimeField.default_timezone` when a custom timezone is set. [#8531]
+* Make relative URLs clickable in Browseable API. [#8464]
+* Support `ManyRelatedField` falling back to the default value when the attribute specified by dot notation doesn't exist. Matches `ManyRelatedField.get_attribute` to `Field.get_attribute`. [#7574]
+* Make `schemas.openapi.get_reference` public. [#7515]
+* Make `ReturnDict` support `dict` union operators on Python 3.9 and later. [#8302]
+* Update throttling to check if `request.user` is set before checking if the user is authenticated. [#8370]
```
Let's make these actual links:

```suggestion
* Enforce `is_valid(raise_exception=False)` as a keyword-only argument. [[#7952](https://github.com/encode/django-rest-framework/pull/7952)]
* Django 4.1 compatability. [[#8591](https://github.com/encode/django-rest-framework/pull/8591)]
* Stop calling `set_context` on Validators. [[#8589](https://github.com/encode/django-rest-framework/pull/8589)]
* Return `NotImplemented` from `ErrorDetails.__ne__`. [[#8538](https://github.com/encode/django-rest-framework/pull/8538)]
* Don't evaluate `DateTimeField.default_timezone` when a custom timezone is set. [[#8531](https://github.com/encode/django-rest-framework/pull/8531)]
* Make relative URLs clickable in Browseable API. [[#8464](https://github.com/encode/django-rest-framework/pull/8464)]
* Support `ManyRelatedField` falling back to the default value when the attribute specified by dot notation doesn't exist. Matches `ManyRelatedField.get_attribute` to `Field.get_attribute`. [[#7574](https://github.com/encode/django-rest-framework/pull/7574)]
* Make `schemas.openapi.get_reference` public. [[#7515](https://github.com/encode/django-rest-framework/pull/7515)]
* Make `ReturnDict` support `dict` union operators on Python 3.9 and later. [[#8302](https://github.com/encode/django-rest-framework/pull/8302)]
* Update throttling to check if `request.user` is set before checking if the user is authenticated. [[#8370](https://github.com/encode/django-rest-framework/pull/8370)]
```

Nta1e commented:
@tomchristie can we release this? It's blocking upgrades to Django 4.1

tomchristie commented:
Consideration... I suppose we might want to include #8617 for this release. Thoughts?

We also ought to add a https://www.django-rest-framework.org/community/3.14-announcement/ page to the docs.

mscansian commented:
@tomchristie it would be nice to not delay too much the release of this version since it's actually blocking everyone from upgrading to Django 4.1

tim-schilling commented:
I can add the community page this weekend. If #8617 isn't resolved by then, perhaps that goes in a 3.15 release when it does get merged?

tim-schilling commented:
I've added a draft for the community page, including the mention for #7522.

I also noticed that the warnings for `RemovedInDRF314Warning` were still being raised, so I removed the functionality that was scheduled for removal.

One thing that I caught was that the openapi `_get_reference` warning was introduced into the master branch, but never released. Since the deprecation is being released in version 3.14, we can't remove the call in 3.14 as well. I've created a new class for `RemovedInDRF315Warning` and switched `_get_reference` to reference that instead.

adamchainz commented on rest_framework/__init__.py:
```diff
@@ -35,3 +35,7 @@ class RemovedInDRF313Warning(DeprecationWarning):
 
 class RemovedInDRF314Warning(PendingDeprecationWarning):
     pass
```
I think this and the above `RemovedInDRF313Warning` can be removed? I cannot find any other references.

tim-schilling replied:
I kept it as per the comments in https://github.com/encode/django-rest-framework/pull/8589.

adamchainz replied:
That argument doesn't really convince me, DRF has never kept around the old warning classes for previous versions, e.g. `RemovedInDRF312Warning` was removed in d45e0005f32bc246a49b209836d233f3d23d77b0. Similarly, Django does not keep around old warning classes "just in case".

tim-schilling replied:
I'd rather leave that decision up to a maintainer. I wasn't a part of the original discussion so it'd be disingenuous for me to say that by convincing me, the past argument is moot.

DavidCain replied:
Sorry to have missed this thread!

I was assuming that `RemovedInDRF313Warning` would be deleted with the 3.14 release.
(My point was only that deleting an exception class in a minor release would constitute a breaking change)

DavidCain replied:
CC @adamchainz - put out a PR to delete the now-unused warnings: https://github.com/encode/django-rest-framework/pull/8664

auvipy commented on docs/community/3.14-announcement.md:
```diff
@@ -0,0 +1,62 @@
+
+Our requirements are now:
+
+* Python 3.6+
```
imho this should be 3.7+ but we need https://github.com/encode/django-rest-framework/pull/8518 to be merged
```suggestion
* Python 3.7+
```

adamchainz replied:
I'd be happy merging that but I'll defer to @tomchristie.

auvipy commented on docs/community/3.14-announcement.md:
```diff
@@ -0,0 +1,62 @@
+Our requirements are now:
+
+* Python 3.6+
+* Django 4.1, 4.0, 3.2, 3.1, 2.2 (LTS)
```
```suggestion
* Django 4.1, 4.0, 3.2 (LTS)
```

auvipy replied:
I can do the related clean ups in a separate PR as well.

spapas commented:
Hello friends, do we have a timeline on when will the new release be released? 

I think this should be prioritized because DRF can't be used with Django 4.1 because of the `ValueError: The file 'rest_framework/css/bootstrap-theme.min.css.map' could not be found with <django.contrib.staticfiles.storage.ManifestStaticFilesStorage` error.

Thank you and kind regards,
Serafeim

spapas commented:
Also, are you working on a specific branch or tag for the 3.14 release ? I'd be happy to install it directly from here  using the `-e 
git+https://github.com...` requirement if you point me to the correct branch. Could I use `tim-schilling:version-3.14.0` ?

tomchristie commented:
> Hello friends, do we have a timeline on when will the new release be released?

It's right at the top of the list. Ought to be this week.

tomchristie commented:
> Ought to be this week.

Arrrgggh. Don't be too mean.
Getting there.

tim-schilling commented:
Tom, what about the comments from the reviews? There were some good points
that we were looking for your opinion on.

On Wed, Sep 21, 2022, 8:08 AM Tom Christie ***@***.***> wrote:

> Merged #8599 <https://github.com/encode/django-rest-framework/pull/8599>
> into master.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/encode/django-rest-framework/pull/8599#event-7430771172>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJYZP2PQUZWAEOT34XDZK3V7MCEVANCNFSM56E5GBCQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

tomchristie commented:
Okay... I'm ready for us to release this unless there's any objections?

tim-schilling commented:
@tomchristie is DRF's policy regarding `RemovedInDRF<version>Warning` changing? Adam brought up that the project hasn't kept the previous warning classes, but we still have `RemovedInDRF314Warning` hanging around currently.

gopackgo90 commented:
It looks like Django 2.2 is being dropped, but the setup.cfg still looks like it allows for Django 2.2 both in the `python_requires` and `classifiers`.

https://github.com/encode/django-rest-framework/blob/master/setup.py#L85
https://github.com/encode/django-rest-framework/blob/master/setup.py#L92

I also would think dropping Django 2.2 should be added to the release notes.

tomchristie commented:
Okay, released.

pauloxnet commented:
Thanks for the release. 🙏️

MarkusH commented:
> Okay, released.

Wonderful. Thank you :heart:
